pkg/api/bzz.go

```diff
@@ -0,0 +1,191 @@
+	_, err = file.JoinReadAll(j, address, buf, toDecrypt)
+	if err != nil {
+		s.Logger.Debugf("bzz download: read entry %s: %v", address, err)
+		s.Logger.Errorf("bzz download: read entry %s", address)
+		jsonhttp.NotFound(w, nil)
+		return
+	}
+	e := &entry.Entry{}
+	err = e.UnmarshalBinary(buf.Bytes())
+	if err != nil {
+		s.Logger.Debugf("bzz download: unmarshal entry %s: %v", address, err)
+		s.Logger.Errorf("bzz download: unmarshal entry %s", address)
+		jsonhttp.InternalServerError(w, "error unmarshaling entry")
+		return
+	}
+
+	// read metadata
+	buf = bytes.NewBuffer(nil)
+	_, err = file.JoinReadAll(j, e.Metadata(), buf, toDecrypt)
+	if err != nil {
+		s.Logger.Debugf("bzz download: read metadata %s: %v", address, err)
+		s.Logger.Errorf("bzz download: read metadata %s", address)
+		jsonhttp.NotFound(w, nil)
+		return
+	}
+	metadata := &entry.Metadata{}
+	err = json.Unmarshal(buf.Bytes(), metadata)
+	if err != nil {
+		s.Logger.Debugf("bzz download: unmarshal metadata %s: %v", address, err)
+		s.Logger.Errorf("bzz download: unmarshal metadata %s", address)
+		jsonhttp.InternalServerError(w, "error unmarshaling metadata")
+		return
+	}
+
+	// we are expecting manifest Mime type here
+	if ManifestContentType != metadata.MimeType {
+		jsonhttp.BadRequest(w, "not manifest")
+		return
+	}
+
+	// read manifest content
+	buf = bytes.NewBuffer(nil)
+	_, err = file.JoinReadAll(j, e.Reference(), buf, toDecrypt)
+	if err != nil {
+		s.Logger.Debugf("bzz download: data join %s: %v", address, err)
+		s.Logger.Errorf("bzz download: data join %s", address)
+		jsonhttp.NotFound(w, nil)
+		return
+	}
+	manifest, err := s.ManifestParser.Parse(buf.Bytes())
+	if err != nil {
+		s.Logger.Debugf("bzz download: unmarshal manifest %s: %v", address, err)
+		s.Logger.Errorf("bzz download: unmarshal manifest %s", address)
+		jsonhttp.InternalServerError(w, "error unmarshaling manifest")
+		return
+	}
+
+	entry, err := manifest.FindEntry(path)
+	if err != nil {
+		jsonhttp.BadRequest(w, "invalid path address")
+		return
+	}
+
+	entryAddress := entry.GetReference()
+
+	toDecryptEntry := len(entryAddress.Bytes()) == (swarm.HashSize + encryption.KeyLength)
+
+	dataSize, err := j.Size(ctx, entryAddress)
+	if err != nil {
+		if errors.Is(err, storage.ErrNotFound) {
+			s.Logger.Debugf("bzz download: not found %s: %v", entryAddress, err)
+			s.Logger.Error("bzz download: not found")
+			jsonhttp.NotFound(w, "not found")
+			return
+		}
+		s.Logger.Debugf("bzz download: invalid root chunk %s: %v", entryAddress, err)
+		s.Logger.Error("bzz download: invalid root chunk")
+		jsonhttp.BadRequest(w, "invalid root chunk")
+		return
+	}
+
+	outBuffer := bytes.NewBuffer(nil)
+	c, err := file.JoinReadAll(j, entryAddress, outBuffer, toDecryptEntry)
+	if err != nil && c == 0 {
+		s.Logger.Debugf("bzz download: data join %s: %v", entryAddress, err)
+		s.Logger.Errorf("bzz download: data join %s", entryAddress)
+		jsonhttp.NotFound(w, nil)
+		return
+	}
+
+	pr, pw := io.Pipe()
```

Comment:
why all this complexity here? this should be the same as downloading a file from here? why no shared code?

Reply:
I agree, we mentioned this already: https://beehive.ethswarm.org/swarm/pl/83bayftmzfgqbqgocpyxwkn3xc. But we decided to c/p for dev speed :) @zbiljic maybe we could extract it as a separate function in `api` if not in the special component right away?